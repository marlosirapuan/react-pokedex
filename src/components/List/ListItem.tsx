import React, { useState, useEffect, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from 'store/rootReducer'

import { add, remove } from 'store/pokedexSlice'
import { PokemonResultItem, Pokemon } from 'typings'
import api from 'api'

import { Box, Text, Button } from '@chakra-ui/react'
import { AiOutlineStar, AiFillStar } from 'react-icons/ai'

import imageFallback from 'assets/image-not-found.png'

type ListItemProps = {
  item: PokemonResultItem
}

const ListItem = ({ item }: ListItemProps): JSX.Element => {
  const dispatch = useDispatch()
  const pokedex = useSelector((state: RootState) => state.pokedex)

  const [pokemonListItem, setPokemonListItem] = useState<Pokemon>({} as Pokemon)

  const getInfoPokemon = async (name: string): Promise<void> => {
    try {
      const { data } = await api.get<Pokemon>(`/pokemon/${name}`)
      setPokemonListItem(data)
    } catch (error) {
      setPokemonListItem({ error: true })
    }
  }

  const isFavorited = (name: string | undefined): boolean => {
    if (!name) {
      return false
    }

    return pokedex.some((pokemon) => pokemon.name === name)
  }

  const handleToggleFavorite = useCallback(
    (pokemon: Pokemon | undefined) => {
      if (pokemon) {
        if (
          pokemon.name &&
          pokedex.some((item) => item.name === pokemon.name)
        ) {
          dispatch(remove(pokemon.name))
        } else {
          dispatch(add(pokemon))
        }
      }
    },
    [dispatch, pokedex]
  )

  useEffect(() => {
    getInfoPokemon(item.name)
  }, [item.name])

  return (
    <Box
      key={pokemonListItem.name}
      display="flex"
      flexDirection="column"
      alignItems="center"
      justifyContent="center"
      minHeight={100}
      borderRadius={10}
      borderWidth={4}
      padding={3}
    >
      <Text fontSize="lg">{pokemonListItem.name}</Text>
      <img
        src={pokemonListItem.sprites?.front_default ?? imageFallback}
        alt={pokemonListItem.name}
      />
      <Button
        variant="ghost"
        _active={{ background: 'none' }}
        _hover={{ background: 'none' }}
        _focus={{ border: 'none' }}
        onClick={() => handleToggleFavorite(pokemonListItem)}
      >
        {isFavorited(pokemonListItem.name) ? (
          <AiFillStar color="orange" size={32} />
        ) : (
          <AiOutlineStar color="orange" size={32} />
        )}
      </Button>
      <Button variant="ghost">Detalhar</Button>
    </Box>
  )
}

export default ListItem
