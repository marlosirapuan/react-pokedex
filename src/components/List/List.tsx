import React, { useState, useEffect, useMemo } from 'react'

import ListItem from './ListItem'
import api from 'api'
import { PokemonResults, PokemonResultItem } from 'typings'

import {
  HStack,
  Skeleton,
  SimpleGrid,
  Alert,
  AlertIcon,
  useToast,
  useBreakpointValue
} from '@chakra-ui/react'

const LIMIT = 2000
const LIMIT_PER_PAGE = 18

type ListProps = {
  filter?: string
}

const List = ({ filter }: ListProps): JSX.Element => {
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState<string | null>(null)
  const [pokemonResult, setPokemonResult] = useState<PokemonResults>(
    {} as PokemonResults
  )

  const toast = useToast()

  const fetchPokemons = async (): Promise<void> => {
    try {
      setLoading(true)
      setError(null)
      const { data } = await api.get<PokemonResults>(`/pokemon?limit=${LIMIT}`)
      setPokemonResult(data)
    } catch (error) {
      setError(error.toString())
    } finally {
      setLoading(false)
    }
  }

  const pokemonResultsMemo = useMemo(() => {
    if (!filter) {
      return pokemonResult.results || []
    }

    return (
      pokemonResult.results?.filter((record) => {
        return record.name.toLowerCase().includes(filter.toLowerCase())
      }) || []
    )
  }, [filter, pokemonResult.results])

  const columns = useBreakpointValue({ base: 2, md: 3 })

  useEffect(() => {
    fetchPokemons()
  }, [])

  useEffect(() => {
    if (!error) {
      return
    }

    toast({
      title: 'Erro',
      description: 'Erro ao fazer requisição. Favor, tente novamente.',
      status: 'error',
      duration: 3000,
      isClosable: true
    })
  }, [error, toast])

  return (
    <div aria-label="list">
      {loading && (
        <HStack spacing={5}>
          <Skeleton height={80} width={80} />
          <Skeleton height={80} width={80} />
          <Skeleton height={80} width={80} />
        </HStack>
      )}

      {!loading && (
        <>
          <SimpleGrid columns={columns} spacing={5}>
            {pokemonResultsMemo
              .slice(0, LIMIT_PER_PAGE)
              .map((item: PokemonResultItem) => {
                return <ListItem key={item.name} item={item} />
              })}
          </SimpleGrid>

          {!pokemonResultsMemo.length && (
            <Alert status="error">
              <AlertIcon />
              Nenhum pokemon encontrado com &quot;{filter}&quot;
            </Alert>
          )}
        </>
      )}
    </div>
  )
}

export default List
