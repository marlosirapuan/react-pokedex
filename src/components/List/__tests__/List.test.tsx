import React from 'react'
import { shallow } from 'enzyme'

import List from '../List'

describe('List component', () => {
  it('should be able to render', () => {
    const listComponent = shallow(<List />)
    expect(listComponent).toMatchSnapshot()
  })
})
