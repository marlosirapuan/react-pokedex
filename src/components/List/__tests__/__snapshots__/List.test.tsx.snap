// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should be able to render 1`] = `ShallowWrapper {}`;
